### PR TITLE
fix: make unit tests mock bindings via typed wrappers to fix recent failures

### PR DESCRIPTION
**Pull request by nolanmar511**
It was necessary to create wrappers for the heap and time profiler native components to stub out these components with sinon. 

Before doing this, there were compliation errors like: 
```
ts/test/test-time-profiler.ts:45:11 - error TS2349: Cannot invoke an expression whose type lacks a call signature. Type '((obj: any) => SinonStub<any[], any>) | ((obj: {}) => SinonStub<unknown[], {}>)' has no compatible call signatures.
```

### CODE REVIEW

**Comment by aalexand** [comment]
@ofrobots Could you take a look at this CL? I am fine to LGTM it but I was wondering if there is a simpler way to fix it - such as could we provide *.d.ts type definitions instead of having wrappers?

File: ts/src/profilers/heap-profiler-wrapper.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AllocationProfileNode} from '../v8-types';
+const profiler = require('bindings')('sampling_heap_profiler');
```
Comment (aalexand) [inline ts/src/profilers/heap-profiler-wrapper.ts]: Having "wrapper" in the file name is a bit ugly and not too informative (it does not communicate much on the purpose of the wrapping).

Could we have bindings.ts with approach like https://stackoverflow.com/a/44467330/848626 suggests?

File: ts/src/v8-types.ts [hunk 2]
```diff
@@ -14,6 +14,8 @@
  * limitations under the License.
  */
 
+const heapProfiler = require('bindings')('sampling_heap_profiler');
```
Comment (aalexand) [inline ts/src/v8-types.ts]: Is this expected to be hear? It's not obvious why this belongs to this file (and why only for the heap profiler, not time).

**Comment by aalexand** [comment]
@nolanmar511 Looks good overall, left one comment.

@ofrobots PTAL as well.

File: ts/src/v8-types.ts (hunk 2 above)
Comment (nolanmar511) [reply ts/src/v8-types.ts]: Thanks. That was an error.

**Comment by ofrobots** [comment]
I am unable to reproduce the error reporting in the OP. @nolanmar511 I would like to understand this a bit better before this lands.

**Comment by nolanmar511** [comment]
I had to remove my package-lock.json from my directory to reproduce, so I don't have my original package-lock.json for comparison. 

There was a new release of `@types/sinon` (5.0.6) two days ago. This roughly correlates to these errors appearing, and when I use an older version of `@types/sinon` there are no compile errors.

Sinon now seems to require that the type definition of the function used to stub matches the original function's type definition.

**Comment by nolanmar511** [comment]
I could maybe file an upstream bug (https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30657 is also related to a problem with the 5.0.6 of `@types/sinon`), but I think doing that would only delay making these changes necessary.

**Comment by nolanmar511** [comment]
Also, I see landing this (or some other fix, like pinning `@types/sinon` to 5.0.5) as high priority, since no unit or integration tests are passing, so all other PRs are blocked by this.

**Comment by ofrobots** [comment]
Over the weekend there was a change in `@types/sinon` that seems to have broken the whole universe. @JustinBeckwith has more details and may already have a tracking bug on definitelytyped repo.

I would recommend pinning `@types/sinon` for now as being done in all the other client library repos.

If always-functional unit-tests is a priority, then we should add a `package-lock.json`.

**Comment by JustinBeckwith** [comment]
So @types/sinon 5.0.7 is out, and should? fix some of these issues.  Can you give that a try?

**Comment by nolanmar511** [comment]
I see the same errors with @types/sinon 5.0.7

I think it's reasonable for sinon to require that the type of the method used for stubbing and the stubbed function match, and that our code should change to accommodate this.

Here's the output:
```
nolanmar@nolanmar:~/cloud-profiler-nodejs-working/cloud-profiler-nodejs$ npm install @types/sinon@latest
npm notice created a lockfile as package-lock.json. You should commit this file.
+ @types/sinon@5.0.7
updated 1 package and audited 1286 packages in 2.177s
found 0 vulnerabilities

nolanmar@nolanmar:~/cloud-profiler-nodejs-working/cloud-profiler-nodejs$ npm run compile

> @google-cloud/profiler@0.2.2 compile /usr/local/google/home/nolanmar/cloud-profiler-nodejs-working/cloud-profiler-nodejs
> tsc -p .

ts/test/test-heap-profiler.ts:50:24 - error TS2349: Cannot invoke an expression whose type lacks a call signature. Type '((obj: any) => SinonStub<any[], any>) | ((obj: {}) => SinonStub<unknown[], {}>)' has no compatible call signatures.

50          profileStub = sinon.stub(v8HeapProfiler, 'getAllocationProfile')
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
51                            .returns(copy(v8HeapProfile));
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ts/test/test-heap-profiler.ts:56:12 - error TS2345: Argument of type '{ external: number; rss: number; heapTotal: number; heapUse: number; }' is not assignable to parameter of type 'MemoryUsage'.
  Object literal may only specify known properties, but 'heapUse' does not exist in type 'MemoryUsage'. Did you mean to write 'heapUsed'?

56            heapUse: 2048,
              ~~~~~~~~~~~~~

ts/test/test-heap-profiler.ts:67:24 - error TS2349: Cannot invoke an expression whose type lacks a call signature. Type '((obj: any) => SinonStub<any[], any>) | ((obj: {}) => SinonStub<unknown[], {}>)' has no compatible call signatures.

67          profileStub = sinon.stub(v8HeapProfiler, 'getAllocationProfile')
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
68                            .returns(copy(v8HeapWithPathProfile));
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ts/test/test-heap-profiler.ts:73:12 - error TS2345: Argument of type '{ external: number; rss: number; heapTotal: number; heapUse: number; }' is not assignable to parameter of type 'MemoryUsage'.
  Object literal may only specify known properties, but 'heapUse' does not exist in type 'MemoryUsage'. Did you mean to write 'heapUsed'?

73            heapUse: 2048,
              ~~~~~~~~~~~~~

ts/test/test-heap-profiler.ts:84:24 - error TS2349: Cannot invoke an expression whose type lacks a call signature. Type '((obj: any) => SinonStub<any[], any>) | ((obj: {}) => SinonStub<unknown[], {}>)' has no compatible call signatures.

84          profileStub = sinon.stub(v8HeapProfiler, 'getAllocationProfile')
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
85                            .returns(copy(v8HeapWithPathProfile));
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ts/test/test-heap-profiler.ts:90:12 - error TS2345: Argument of type '{ external: number; rss: number; heapTotal: number; heapUse: number; }' is not assignable to parameter of type 'MemoryUsage'.
  Object literal may only specify known properties, but 'heapUse' does not exist in type 'MemoryUsage'. Did you mean to write 'heapUsed'?

90            heapUse: 2048,
              ~~~~~~~~~~~~~

ts/test/test-profiler.ts:279:34 - error TS2345: Argument of type 'Promise<{}>' is not assignable to parameter of type 'void'.

279                         .returns(new Promise(resolve => {
                                     ~~~~~~~~~~~~~~~~~~~~~~~~
280                           resolve([{}, {statusCode: 200}]);
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
281                         }));
    ~~~~~~~~~~~~~~~~~~~~~~~~~~

ts/test/test-profiler.ts:307:34 - error TS2345: Argument of type 'Promise<{}>' is not assignable to parameter of type 'void'.

307                         .returns(new Promise(resolve => {
                                     ~~~~~~~~~~~~~~~~~~~~~~~~
308                           resolve([{}, {statusCode: 200}]);
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
309                         }));
    ~~~~~~~~~~~~~~~~~~~~~~~~~~

ts/test/test-profiler.ts:332:34 - error TS2345: Argument of type 'Promise<{}>' is not assignable to parameter of type 'void'.

332                         .returns(new Promise(resolve => {
                                     ~~~~~~~~~~~~~~~~~~~~~~~~
333                           resolve([{}, {}]);
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
334                         }));
    ~~~~~~~~~~~~~~~~~~~~~~~~~~

ts/test/test-profiler.ts:361:24 - error TS2345: Argument of type 'Promise<{}>' is not assignable to parameter of type 'void'.

361               .returns(new Promise(resolve => {
                           ~~~~~~~~~~~~~~~~~~~~~~~~
362                 resolve(
    ~~~~~~~~~~~~~~~~~~~~~~~~
363                     [undefined, {statusCode: 500, statusMessage: 'Error 500'}]);
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
364               }));
    ~~~~~~~~~~~~~~~~

ts/test/test-time-profiler.ts:45:11 - error TS2349: Cannot invoke an expression whose type lacks a call signature. Type '((obj: any) => SinonStub<any[], any>) | ((obj: {}) => SinonStub<unknown[], {}>)' has no compatible call signatures.

45           sinon.stub(v8TimeProfiler, 'stopProfiling').returns(v8TimeProfile));
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @google-cloud/profiler@0.2.2 compile: `tsc -p .`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @google-cloud/profiler@0.2.2 compile script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /usr/local/google/home/nolanmar/.npm/_logs/2018-11-20T18_46_54_272Z-debug.log
```

**Comment by JustinBeckwith** [comment]
Agreed.  The fix we made for 5.0.7 was to accept parameters in the `callsFake` method.  So now it's just way more type safe.  which is good :)

**Comment by nolanmar511** [comment]
PTAL

File: ts/src/profilers/heap-profiler-bindings.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AllocationProfileNode} from '../v8-types';
+const profiler = require('bindings')('sampling_heap_profiler');
```
Comment (JustinBeckwith) [inline ts/src/profilers/heap-profiler-bindings.ts]: Can this be an import?

File: ts/test/test-profiler.ts [hunk 4]
```diff
@@ -276,9 +276,7 @@ describe('Profiler', () => {
 
       requestStub = sinon.stub(common.ServiceObject.prototype, 'request')
                         .onCall(0)
-                        .returns(new Promise(resolve => {
-                          resolve([{}, {statusCode: 200}]);
-                        }));
+                        .callsArgWith(1, null, {}, {statusCode: 200});
```
Comment (JustinBeckwith) [inline ts/test/test-profiler.ts]: This is great, but if you wanted to keep the promise workflow you could have used `.resolves(...)`

**Pull request merged**